src/proto/batch/mod.rs

```diff
@@ -0,0 +1,428 @@
+        self
+    }
+
+    /// Create a `Batch` with only `success` callback specified.
+    pub fn with_success_callback(self, success_cb: Job) -> Batch {
+        let mut b = self.build();
+        b.success = Some(success_cb);
+        b
+    }
+
+    /// Create a `Batch` with only `complete` callback specified.
+    pub fn with_complete_callback(self, complete_cb: Job) -> Batch {
+        let mut b = self.build();
+        b.complete = Some(complete_cb);
+        b
+    }
+
+    /// Create a `Batch` with both `success` and `complete` callbacks specified.
+    pub fn with_callbacks(self, success_cb: Job, complete_cb: Job) -> Batch {
+        let mut b = self.build();
+        b.success = Some(success_cb);
+        b.complete = Some(complete_cb);
+        b
+    }
+}
+
+impl Clone for BatchBuilder {
+    fn clone(&self) -> Self {
+        BatchBuilder {
+            parent_bid: self.parent_bid.clone(),
+            description: self.description.clone(),
+            success: self.success.clone(),
+            complete: self.complete.clone(),
+        }
+    }
+}
+
+/// Represents a newly started or re-opened batch of jobs.
+pub struct BatchHandle<'a, S: Read + Write> {
+    bid: String,
+    prod: &'a mut Producer<S>,
+}
+
+impl<'a, S: Read + Write> BatchHandle<'a, S> {
+    /// ID issued by the Faktory server to this batch.
+    pub fn id(&self) -> &str {
+        self.bid.as_ref()
+    }
+
+    pub(crate) fn new(bid: String, prod: &mut Producer<S>) -> BatchHandle<'_, S> {
+        BatchHandle { bid, prod }
+    }
+
+    /// Add the given job to the batch.
+    ///
+    /// Should the submitted job - for whatever reason - already have a `bid` key present in its custom hash,
+    /// this value will be overwritten by the ID of the batch this job is being added to with the old value
+    /// returned as `Some(<old value here>)`.
+    pub fn add(&mut self, mut job: Job) -> Result<Option<serde_json::Value>, Error> {
+        let bid = job.custom.insert("bid".into(), self.bid.clone().into());
+        self.prod.enqueue(job).map(|_| bid)
+    }
+
+    /// Initiate a child batch of jobs.
+    pub fn start_batch(&mut self, mut batch: Batch) -> Result<BatchHandle<'_, S>, Error> {
+        batch.parent_bid = Some(self.bid.clone());
+        self.prod.start_batch(batch)
+    }
+
+    /// Commit this batch.
+    ///
+    /// The Faktory server will not queue any callbacks, unless the batch is committed.
+    /// Committing an empty batch will make the server queue the callback(s) right away.
+    /// Once committed, the batch can still be re-opened with [open_batch](Producer::open_batch),
+    /// and extra jobs can be added to it.
+    pub fn commit(self) -> Result<(), Error> {
+        self.prod.commit_batch(self.bid)
+    }
+}
+
+// Not documented, but existing de fakto and also mentioned in the official client
+// https://github.com/contribsys/faktory/blob/main/client/batch.go#L17-L19
+/// State of a `callback` job of a [`Batch`].
```

Comment:
```suggestion
/// State of a `callback` job of a [`Batch`].
///
// Not documented, but existing de fakto and also mentioned in the official client
// https://github.com/contribsys/faktory/blob/main/client/batch.go#L17-L19
```
rustdoc expects the first paragraph of a doc comment to be the main "tagline" for the item being documented, and will often show _just_ that.

Reply:
Those two lines were not meant to be published.
![image](https://github.com/jonhoo/faktory-rs/assets/117771945/59a8a570-17c3-4c0c-abc1-133a54c635d3)

With the suggested changes they will still be simple comments, but it kind of looks confusing?

Reply:
Ohh, I missed that they were normalt comments! Fine to leave them as is then 👍